### PR TITLE
[OSDEV-2542] Add Spotlight Data Partners contributor filter to platform search

### PR DESCRIPTION
**Pull request by vlad-shapik**
Add Spotlight Data Partners contributor-based filtering to platform search, with grouped options and consistent backend/frontend filter semantics.

Jira: [OSDEV-2542](https://opensupplyhub.atlassian.net/browse/OSDEV-2542)

## Why

Users need to find facilities based on which Spotlight data partners contributed data. The implementation uses partner contributors so the UI model and backend filtering behavior stay aligned, and caches partner-group option responses to reduce repeated API/origin load.

### Backend (Django API)
- Added `partner_contributor` filtering to facility search and added `combine_partner_contributors` to support OR/AND combination semantics.
- Added shared partner-contributor filtering helper logic and reused it across both facility managers.
- Added `GET /api/partner-group-contributors/` to return grouped contributor options for the Spotlight Data Partners filter.
- Scoped filtering to active partner fields that are assigned to groups.
- Updated facilities query parameter/OpenAPI declarations for the contributor-based flow.
- Aligned `PartnerGroupContributorsView` module docs with actual default API permissions.
- Applied view caching to `PartnerGroupContributorsView` via class-level `dispatch` decorator.

### Frontend (React)
- Implemented `DataPartnersFilter` for the **Spotlight Data Partners** UI filter.
- Added partner-group-contributors Redux flow (actions/reducer/selectors) with lazy option loading.
- Added query-string round-trip support for `partner_contributor` and `combine_partner_contributors`.
- Kept map/tile queries aligned with selected Spotlight Data Partners filters.
- Reduced redundant filter-option fetches between homepage and facilities views.

### Infrastructure (CloudFront / Terraform)
- Added dedicated CloudFront cache behavior for `api/partner-group-contributors*`.
- Added configurable TTL variables:
  - `api_partner_group_contributors_cache_default_ttl`
  - `api_partner_group_contributors_cache_max_ttl`

### Tests and docs
- Added backend coverage for partner-contributor OR/AND semantics and edge cases.
- Added API tests for `partner-group-contributors` endpoint behavior.
- Added frontend tests for `DataPartnersFilter` interactions and combine mode behavior.
- Updated `doc/release/RELEASE-NOTES.md` with `OSDEV-2542` entries.

## Out of scope
- No OpenSearch/Logstash changes.
- No `/api/v1/production-locations/` changes.
- No database migrations.

[OSDEV-2542]: https://opensupplyhub.atlassian.net/browse/OSDEV-2542?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/django/api/models/facility/facility_manager.py [hunk 1]
```diff
@@ -202,6 +205,54 @@ def filter_by_query_params(self, params):
...
+        if partner_contributors:
+            from api.models.partner_field import PartnerField
+
+            partner_fields = list(
+                PartnerField.objects.filter(
+                    contributor__id__in=partner_contributors,
+                    active=True,
+                    group__isnull=False,
+                ).values_list('contributor__id', 'name').distinct()
+            )
```
Comment (protsack-stephan) [inline src/django/api/models/facility/facility_manager.py]: I think there's a method in the doebase to retrieve those from cache, can you take a look? That should lead to a better performance, since they are being used in the different pats of the application.

File: src/django/api/models/facility/facility_manager.py [hunk 2]
```diff
@@ -202,6 +205,54 @@ def filter_by_query_params(self, params):
...
+                            _partner_contributor_id,
+                            field_name,
+                        ) in partner_fields
+                    }
+                )
+                if field_names:
+                    facilities_qs = apply_partner_fields_or_filter(
+                        facilities_qs, field_names
+                    )
+
```
Comment (protsack-stephan) [inline src/django/api/models/facility/facility_manager.py]: Can we move this to a separate method? Also, it would be probably a bit easier to read if we were to use more early returns and negative if statements. Here's an example that I think might work:

```python
class FacilityManager(models.Manager):
    def _apply_partner_contributor_filters(self, facilities_qs, params):
        partner_contributors = params.getlist(
            FacilitiesQueryParams.PARTNER_CONTRIBUTOR
        )

        if not partner_contributors:
            return facilities_qs

        combine_partner_contributors = params.get(
            FacilitiesQueryParams.COMBINE_PARTNER_CONTRIBUTORS, ""
        )

        from api.models.partner_field import PartnerField

        partner_fields = (
            PartnerField.objects.filter(
                contributor__id__in=partner_contributors,
                active=True,
                group__isnull=False,
            )
            .values_list("contributor__id", "name")
            .distinct()
        )

        contributor_fields_map = {}

        for contributor_id, field_name in partner_fields:
            contributor_id = str(contributor_id)

            contributor_fields_map.setdefault(contributor_id, set()).add(
                field_name
            )

        if combine_partner_contributors.upper() == "AND":
            for contributor_id in partner_contributors:
                contributor_field_names = list(
                    contributor_fields_map.get(str(contributor_id), set())
                )
                if not contributor_field_names:
                    return facilities_qs.none()
                facilities_qs = apply_partner_fields_or_filter(
                    facilities_qs, contributor_field_names
                )

            return facilities_qs

        field_names = list(
            {
                field_name
                for contributor_field_names in contributor_fields_map.values()
                for field_name in contributor_field_names
            }
        )

        if field_names:
            facilities_qs = apply_partner_fields_or_filter(
                facilities_qs, field_names
            )

        return facilities_qs

    def filter_by_query_params(self, params):
        # other code goes here...
        facilities_qs = self._apply_partner_contributor_filters(
            facilities_qs, params
        )
        # other code goes here...
        return facilities_qs
```

File: src/django/api/models/facility/facility_manager_index_new.py [hunk 3]
```diff
@@ -201,4 +204,52 @@ def filter_by_query_params(self, params):
...
+        if partner_contributors:
+            from api.models.partner_field import PartnerField
+
+            partner_fields = list(
+                PartnerField.objects.filter(
+                    contributor__id__in=partner_contributors,
+                    active=True,
+                    group__isnull=False,
+                ).values_list('contributor__id', 'name').distinct()
+            )
```
Comment (protsack-stephan) [inline src/django/api/models/facility/facility_manager_index_new.py]: Can we get this from cache? I think there's a method that we can use to do that to avoid additional database queries called `get_cached_all_partner_fields`.

File: src/django/api/models/facility/facility_manager_index_new.py [hunk 4]
```diff
@@ -201,4 +204,52 @@ def filter_by_query_params(self, params):
...
+                        for (
+                            _partner_contributor_id,
+                            field_name,
+                        ) in partner_fields
+                    }
+                )
+                if field_names:
+                    facilities_qs = apply_partner_fields_or_filter(
+                        facilities_qs, field_names
+                    )
```
Comment (protsack-stephan) [inline src/django/api/models/facility/facility_manager_index_new.py]: It looks like `AND` is used in the multiple functions, maybe it make sense to move it into a enum? Also can we make something similar here, where we move this to a separate private function within the class? Might be easier to manage, here's an example of what I'm thinking about:

```python
class FacilityIndexNewManager(models.Manager):
    def _apply_partner_contributor_filters(self, facilities_qs, params):
        partner_contributors = params.getlist(
            FacilitiesQueryParams.PARTNER_CONTRIBUTOR
        )
        if not partner_contributors:
            return facilities_qs

        combine_partner_contributors = params.get(
            FacilitiesQueryParams.COMBINE_PARTNER_CONTRIBUTORS, ''
        )

        from api.models.partner_field import PartnerField

        partner_fields = PartnerField.objects.filter(
            contributor__id__in=partner_contributors,
            active=True,
            group__isnull=False,
        ).values_list('contributor__id', 'name').distinct()

        contributor_fields_map = {}
        for contributor_id, field_name in partner_fields:
            contributor_id = str(contributor_id)
            contributor_fields_map.setdefault(contributor_id, set()).add(
                field_name
            )

        if combine_partner_contributors.upper() == 'AND':
            for contributor_id in partner_contributors:
                contributor_field_names = list(
                    contributor_fields_map.get(str(contributor_id), set())
                )
                if not contributor_field_names:
                    return facilities_qs.none()
                facilities_qs = apply_partner_fields_or_filter(
                    facilities_qs, contributor_field_names
                )
            return facilities_qs

        field_names = list(
            {
                field_name
                for contributor_field_names in contributor_fields_map.values()
                for field_name in contributor_field_names
            }
        )
        if field_names:
            facilities_qs = apply_partner_fields_or_filter(
                facilities_qs, field_names
            )
        return facilities_qs

    def filter_by_query_params(self, params):
        #... other code goes here...
        facilities_qs = self._apply_partner_contributor_filters(
            facilities_qs, params
        )

        return facilities_qs
```

File: src/django/api/views/partner_group_contributors/partner_group_contributors_view.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+from api.models.partner_field import PartnerField
+from api.models.partner_field_group import PartnerFieldGroup
+from api.pagination import PartnerFieldGroupCursorPagination
+from api.serializers.partner_group_contributors\
+    .partner_group_contributors_serializer import (
+        PartnerGroupContributorsSerializer,
+    )
+
+
+class PartnerGroupContributorsView(ListAPIView):
```
Comment (protsack-stephan) [inline src/django/api/views/partner_group_contributors/partner_group_contributors_view.py:35]: I would suggest we add caching here, it's one more request to execute on the home page, might benefit from being served from cache even through it's not a complex query, just to make sure database is not busy resolving it every time someone visits the home page.

**Review by protsack-stephan — COMMENTED** [review]
Good work! Couple of comments from me!

File: src/django/api/models/facility/facility_manager.py [hunk 6]
```diff
@@ -202,6 +205,54 @@ def filter_by_query_params(self, params):
...
+                            partner_contributor_id,
+                            field_name,
+                        ) in partner_fields
+                        if str(partner_contributor_id) == str(contributor_id)
+                    ]
+                    if not contributor_field_names:
+                        return facilities_qs.none()
+                    facilities_qs = apply_partner_fields_or_filter(
+                        facilities_qs, contributor_field_names
+                    )
```
[the remark this answers is not in the document]
Comment (VadimKovalenkoSNF) [reply src/django/api/models/facility/facility_manager.py]: @vlad-shapik , can you check this?

**Review by VadimKovalenkoSNF — CHANGES_REQUESTED** [review]
Good work! Can you check one request change from the coderabbitai?

File: src/django/api/models/facility/facility_manager.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vlad-shapik) [reply src/django/api/models/facility/facility_manager.py]: @coderabbitai implemented. The early-return path is removed via shared helper logic; in AND mode with unmapped contributor we now set queryset to `.none()` and continue, preserving the `Facility` return model conversion path in `FacilityManager`.

File: src/django/api/models/facility/facility_manager.py [hunk 7]
```diff
@@ -202,6 +205,54 @@ def filter_by_query_params(self, params):
...
+                        for (
+                            _partner_contributor_id,
+                            field_name,
+                        ) in partner_fields
+                    }
+                )
+                if field_names:
+                    facilities_qs = apply_partner_fields_or_filter(
+                        facilities_qs, field_names
+                    )
```
[the remark this answers is not in the document]
Comment (vlad-shapik) [reply src/django/api/models/facility/facility_manager.py]: @coderabbitai implemented. I extracted shared logic into `apply_partner_contributors_filter` in `src/django/api/models/facility/partner_contributor_filter.py` and replaced duplicated blocks in both facility managers with helper calls.

File: src/django/api/views/partner_group_contributors/partner_group_contributors_view.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            Prefetch('contributor_set', queryset=contributor_qs),
+        )
+
+        return PartnerFieldGroup.objects.prefetch_related(
+            Prefetch(
+                'partner_fields',
+                queryset=active_fields_qs,
+                to_attr='active_fields',
+            )
+        )
```
[the remark this answers is not in the document]
Comment (vlad-shapik) [reply src/django/api/views/partner_group_contributors/partner_group_contributors_view.py:83]: @coderabbitai addressed the docstring mismatch by updating the view module docs to state it uses default API permissions (`IsAuthenticatedOrWebClient`). I did not add `Contributor.active` filtering because `Contributor` has no `active` field in this codebase.

File: src/django/api/pagination.py [hunk 9]
```diff
@@ -14,3 +14,16 @@ class FacilitiesGeoJSONPagination(GeoJsonPagination):
...
+    """
+    Cursor based pagination for partner field groups.
+    Allows the client to control the page size via the ?limit= parameter.
+    And adds the default ordering by the `order` field.
+    """
+
+    page_size = 20
+    ordering = "order"
+    page_size_query_param = "limit"
+    max_page_size = 100
```
[the remark this answers is not in the document]
Comment (vlad-shapik) [reply src/django/api/pagination.py:29]: @coderabbitai implemented. Updated cursor pagination ordering to deterministic tuple `("order", "uuid")` in `src/django/api/pagination.py` to avoid tie instability.

File: src/django/api/models/facility/facility_manager.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vlad-shapik) [reply src/django/api/models/facility/facility_manager.py]: @coderabbitai follow-up complete on the review items in this thread; thanks.

**Comment by vlad-shapik** [comment]
@coderabbitai all outstanding inline suggestions from your review were triaged and addressed on branch `OSDEV-2542-add-partner-field-group-and-data-partner-filters` in commit `a2bfb1f4`.

File: src/django/api/models/facility/facility_manager.py (hunk 2 above)
Comment (vlad-shapik) [reply src/django/api/models/facility/facility_manager.py]: Done

File: src/django/api/models/facility/facility_manager_index_new.py (hunk 4 above)
Comment (vlad-shapik) [reply src/django/api/models/facility/facility_manager_index_new.py]: It was moved into a reusable helper function.

File: src/django/api/models/facility/facility_manager_index_new.py (hunk 3 above)
Comment (vlad-shapik) [reply src/django/api/models/facility/facility_manager_index_new.py]: Checked this as well. `get_cached_all_partner_fields` is not a drop-in replacement for this manager filter because it does not carry contributor linkage needed for `partner_contributor` semantics. I kept the current query approach in shared helper to preserve correctness and avoid introducing extra relation lookups here without dedicated cached mapping.

File: src/django/api/models/facility/facility_manager.py (hunk 1 above)
Comment (vlad-shapik) [reply src/django/api/models/facility/facility_manager.py]: Thanks Stephan. I looked into cache reuse here and kept the current DB query path for partner contributor filtering. The existing `get_cached_all_partner_fields` cache stores only PartnerField rows; this filter needs contributor->field mapping and using that helper directly would require additional relation fetches (or a new dedicated cached mapping) to avoid N+1 patterns. Given scope, I left this as-is.

File: src/django/api/views/partner_group_contributors/partner_group_contributors_view.py (hunk 5 above)
Comment (vlad-shapik) [reply src/django/api/views/partner_group_contributors/partner_group_contributors_view.py:35]: Implemented. I added endpoint caching at CloudFront level for `api/partner-group-contributors*` in `deployment/terraform/cdn.tf` with dedicated TTL variables in `deployment/terraform/variables.tf`. Also switched Django view caching to class-level dispatch decorator for consistency with existing CBV patterns.

**Review by daniellekain — DISMISSED** [review]
thanks for making all of theses changes! I focused on the updates you've made since this was last approved.

**Review by protsack-stephan — APPROVED** [review]

File: src/django/api/models/facility/facility_manager.py (hunk 1 above)
Comment (protsack-stephan) [reply src/django/api/models/facility/facility_manager.py]: Sounds good. Thanks for checking.

File: src/django/api/models/facility/facility_manager_index_new.py (hunk 3 above)
Comment (protsack-stephan) [reply src/django/api/models/facility/facility_manager_index_new.py]: Sounds good. Thanks for checking.

**Review by daniellekain — APPROVED** [review]

**Review by VadimKovalenkoSNF — APPROVED** [review]
